torchgeo/datamodules/vaihingen.py

```diff
@@ -63,24 +87,77 @@ def setup(self, stage: Optional[str] = None) -> None:
 
         Args:
             stage: stage to set up
+
+        .. versionchanged:: 0.4
+            Add functionality to randomly crop patches from a tile during
+            training and pad validation and test samples to next multiple of 32
         """
-        transforms = Compose([self.preprocess])
 
-        dataset = Vaihingen2D(split="train", transforms=transforms, **self.kwargs)
+        def n_random_crop(sample: Dict[str, Any]) -> Dict[str, Any]:
+            """Construct 'num_patches_per_tile' random patches of input tile.
+
+            Args:
+                sample: contains image and mask tile from dataset
+
+            Returns:
+                stacked randomly cropped patches from input tile
+            """
+            images, masks = [], []
+            for i in range(self.num_patches_per_tile):
+                mask = repeat(sample["mask"], "h w -> t h w", t=2).float()
+                image, mask = self.rcrop(sample["image"], mask)
+                mask = mask.squeeze()[0]
```

Comment:
There's a lot of squeezing going on here. It's better to specify which singleton dimension your are removing. I think if you don't squeeze then you can replace torch.stack with torch.cat.

Reply:
Yes, this is really important so that `batch_size=1` works correctly (often used in our unit tests)

Reply:
I tested the last commit with `batch_size=1` and `num_patches_per_tile=1`